crates/editor/src/clangd_ext.rs

```diff
@@ -67,9 +67,18 @@ pub fn switch_source_header(
             )
         })?;
 
+
+        let path = if cfg!(windows) {
+            // Url::parse() will leave one forward slash at the start of the string.
```

Comment:
https://en.wikipedia.org/wiki/File_URI_scheme (for simplicity, sure we can ascend up to the corresponding RFCs) denotes the URI format as `file://host/path`

When I apply the
```diff
diff --git a/crates/editor/src/clangd_ext.rs b/crates/editor/src/clangd_ext.rs
index 501f81b107..6fb2867d46 100644
--- a/crates/editor/src/clangd_ext.rs
+++ b/crates/editor/src/clangd_ext.rs
@@ -60,12 +60,13 @@ pub fn switch_source_header(
             return Ok(());
         }

-        let goto = Url::parse(&switch_source_header.0).with_context( {
+        let goto = Url::parse(dbg!(&switch_source_header.0)).with_context( {
             format!(
                 "Parsing URL \"{}\" returned from switch source/header failed",
                 switch_source_header.0
             )
         })?;
+        dbg!(&goto, goto.path());

         workspace
             .update(&mut cx, |workspace, view_cx| {
```

patch and run Zed, I see this:

```
[crates\editor\src\clangd_ext.rs:63:31] &switch_source_header.0 = "file:///C:/work/zed-20583/MyClass.h"
[crates\editor\src\clangd_ext.rs:69:9] &goto = Url {
    scheme: "file",
    cannot_be_a_base: false,
    username: "",
    password: None,
    host: None,
    port: None,
    path: "/C:/work/zed-20583/MyClass.h",
    query: None,
    fragment: None,
}
[crates\editor\src\clangd_ext.rs:69:9] goto.path() = "/C:/work/zed-20583/MyClass.h"
```

So, I think what we do here is merely a hack and we do not fix the underlying issue.

When I apply another patch:
```diff
diff --git a/crates/project/src/lsp_ext_command.rs b/crates/project/src/lsp_ext_command.rs
index 7890630e31..8e01c005a1 100644
--- a/crates/project/src/lsp_ext_command.rs
+++ b/crates/project/src/lsp_ext_command.rs
@@ -293,6 +293,7 @@ impl LspCommand for SwitchSourceHeader {
         _: &Arc<LanguageServer>,
         _: &AppContext,
     ) -> SwitchSourceHeaderParams {
+        dbg!(path, lsp::Url::from_file_path(path));
         SwitchSourceHeaderParams(lsp::TextDocumentIdentifier {
             uri: lsp::Url::from_file_path(path).unwrap(),
         })
```

I see that a completely valid Windows abs path was turned into something odd:
```
[crates\project\src\lsp_ext_command.rs:296:9] path = "C:\\work\\zed-20583\\MyClass.cpp"
[crates\project\src\lsp_ext_command.rs:296:9] lsp::Url::from_file_path(path) = Ok(
    Url {
        scheme: "file",
        cannot_be_a_base: false,
        username: "",
        password: None,
        host: None,
        port: None,
        path: "/C:/work/zed-20583/MyClass.cpp",
        query: None,
        fragment: None,
    },
)
```

I am not good with Windows internals, but that seems incorrect?
It either should be a UNC path, or something that's not starting with `/` at all.

Reply:
<details>
My patch is definitely hack-ish. My rationale was that everything I threw at Zed so far worked (drag-n-drop, projects, search, etc) meaning fs for Windows was functional. This way - to me - this is here where it was wrong and so I decided to patch it here. 

I am obviously not attached to this solution and open to fixing it for real instead of hacking along. I PR'd this patch mostly because I wanted to make progress for myself (I use switch header source a lot) and get feedback from maintainers.

> I am not good with Windows internals, but that seems incorrect?

You mean forward slashes instead of backslashes? To my limited knowledge of Windows internals *I think* this should work. If this doesn't natively work, fs abstraction layer should transform path separators under the hood.
</details>

Reply:
I've skimmed url library docs and it seems they already have what we want? https://docs.rs/url/2.5.2/url/struct.Url.html#method.to_file_path. In the docs they even have the example which basically represents what we do already:

```rust
    /// # use url::Url;
    /// # let url = Url::parse("file:///etc/passwd").unwrap();
    /// let path = url.to_file_path();
```

On my machine it did give correct path and I was able to pass it straight up to `workspace::open_abs_path()`.

Reply:
Beautiful!
So we have to replace `.path()` with that call instead and that's a proper fix?
If it works for you, let's do that instead (and, optionally, we can replace other usages in a separate commit, but not insisting).

Reply:
Updated the patch and left a review-comment asking if this way of handling error is fine. 

---

> optionally, we can replace other usages in a separate commit, but not insisting

There are 2 other usages of `Url::path()` (at least that's what my IDE tells me).

https://github.com/zed-industries/zed/blob/main/crates/zed/src/zed/open_listener.rs#L96

This one seems fine and no need to change.

https://github.com/zed-industries/zed/blob/main/crates/git/src/remote.rs#L70

This is a test. It would be reasonable to also add Windows style path here, but I am not sure if I should.  I think fix here would also require introduction of wrapper function to get either `Url::path()` or `Url::to_file_name()` depending on schema. On the other hand, this is only used in a test and nowhere else.

Reply:
Thank you.

Not sure why the first is fine, but both seem quite non-significant and can be tackled later (as we need to enable tests on Windows anyway, and the first one is sort of guarded by the parsing logic so we'll see it if acts up)